Incorrect Use of `.offset_from(1)` When Transpiling Pointer Decrement on Struct Field

## Description

C2Rust incorrectly translates valid C code containing a pointer decrement on a volatile struct field, such as` --(global.p);`, into Rust using `.offset_from(1)`, which is invalid and causes a type mismatch at compile time.

## C code:

```c
int arr[10];
struct S { int * volatile p; } global;
int func(int *restrict r) {
    --(global.p);
    return *r;
}

int main() {
     return 0;
}
```

## Translated Rust Code

```rust
#[derive(Copy, Clone)]
#[repr(C)]
pub struct S {
    pub p: *mut libc::c_int,
}
#[no_mangle]
pub static mut arr: [libc::c_int; 10] = [0; 10];
#[no_mangle]
pub static mut global: S = S {
    p: 0 as *const libc::c_int as *mut libc::c_int,
};
#[no_mangle]
pub unsafe extern "C" fn func(mut r: *mut libc::c_int) -> libc::c_int {
    ::core::ptr::write_volatile(
        &mut global.p as *mut *mut libc::c_int,
        (::core::ptr::read_volatile::<
            *mut libc::c_int,
        >(&global.p as *const *mut libc::c_int))
            .offset_from(1) as *mut libc::c_int,
    );
    ::core::ptr::read_volatile::<*mut libc::c_int>(&global.p as *const *mut libc::c_int);
    return *r;
}
```

## Cargo build error:

```text
error[E0308]: mismatched types
  --> src/runner.rs:29:26
   |
29 |             .offset_from(1) as *mut libc::c_int,
   |              ----------- ^ expected `*const i32`, found `usize`
   |              |
   |              arguments to this method are incorrect
   |
   = note: expected raw pointer `*const i32`
                     found type `usize`
note: method defined here
  --> /rustc/05f9846f893b09a1be1fc8560e33fc3c815cfecb/library/core/src/ptr/mut_ptr.rs:856:25
```


c2rust version: `C2Rust 0.20.0`

--- a/c2rust-transpile/src/translator/operators.rs
+++ b/c2rust-transpile/src/translator/operators.rs
@@ -727,6 +727,12 @@ impl<'c> Translation<'c> {
         } else {
             c_ast::BinOp::AssignSubtract
         };
+
+        let arg_type = self.ast_context[arg]
+            .kind
+            .get_qual_type()
+            .ok_or_else(|| format_err!("bad arg type"))?;
+
         let one = match self.ast_context.resolve_type(ty.ctype).kind {
             // TODO: If rust gets f16 support:
             // CTypeKind::Half |
@@ -741,19 +747,27 @@ impl<'c> Translation<'c> {
             }
             _ => mk().lit_expr(mk().int_unsuffixed_lit(1)),
         };
-        let arg_type = self.ast_context[arg]
-            .kind
-            .get_qual_type()
-            .ok_or_else(|| format_err!("bad arg type"))?;
+
+        let one_type_id =
+            if let CTypeKind::Pointer(..) = self.ast_context.resolve_type(arg_type.ctype).kind {
+                CQualTypeId::new(
+                    self.ast_context
+                        .type_for_kind(&CTypeKind::Int)
+                        .ok_or_else(|| format_err!("couldn't find type for CTypeKind::Int"))?,
+                )
+            } else {
+                arg_type
+            };
+
         self.convert_assignment_operator_with_rhs(
             ctx.used(),
             op,
-            arg_type,
-            arg,
             ty,
+            arg,
+            one_type_id,
             WithStmts::new_val(one),
             Some(arg_type),
-            Some(arg_type),
+            Some(ty),
         )
     }
 

--- a/c2rust-transpile/tests/snapshots.rs
+++ b/c2rust-transpile/tests/snapshots.rs
@@ -410,6 +410,11 @@ fn test_str_init() {
     transpile("str_init.c").run();
 }
 
+#[test]
+fn test_volatile() {
+    transpile("volatile.c").expect_compile_error(true).run();
+}
+
 // arch-specific
 
 #[test]

--- a/c2rust-transpile/tests/snapshots.rs
+++ b/c2rust-transpile/tests/snapshots.rs
@@ -412,7 +412,10 @@ fn test_str_init() {
 
 #[test]
 fn test_volatile() {
-    transpile("volatile.c").expect_compile_error(true).run();
+    transpile("volatile.c")
+        // https://github.com/immunant/c2rust/pull/1689#discussion_r3015140974
+        .expect_compile_error_edition_2024(true)
+        .run();
 }
 
 // arch-specific

--- a/c2rust-transpile/tests/snapshots/volatile.c
+++ b/c2rust-transpile/tests/snapshots/volatile.c
@@ -1,0 +1,6 @@
+struct S { int * volatile p; } volatile_global_struct;
+
+void test_volatile(void) {
+    // https://github.com/immunant/c2rust/issues/1237
+    --(volatile_global_struct.p);
+}

